### DISCUSSION TITLE
Support of Mysql UNSIGNED DOUBLE

### DISCUSSION THREAD

**Discussion by zim32** [diesel-rs/diesel#3749] [Ideas]
Diesel should support Mysql UNSIGNED DOUBLE type, because this is just DOUBLE which can not be negative.

From Mysql docs:

`UNSIGNED, if specified, disallows negative values. As of MySQL 8.0.17, the UNSIGNED attribute is deprecated for columns of type DOUBLE (and any synonyms) and you should expect support for it to be removed in a future version of MySQL. Consider using a simple CHECK constraint instead for such columns.`

With this being said, it means that it is safe to map unsigned double into f64, but reverse mapping will not be safe. Actually to be able to provide full safety diesel should support parsing and checking mysql CHECK constraints, which is not really possible I guess. 

So I propopse to just add mapping of Unsigned(Double) to f64.

**Comment by weiznich** [comment]
We have the following policy for type mappings:

* It's a commonly used type
* We can provide `FromSql` and `ToSql` implementations 
* The conversion cannot fail for all possible values of that type.

(There are some exceptions for this for types universally used…)

The first two points are arguably covered by this proposal. The third point is not covered as inserting a negative `f64` value would fail. I personally do not consider this type mapping to be that commonly used so that it would justify an exception from the rules listed above.

**Comment by zim32** [comment]
First of all I want to mention that right now diesel is not fully type safe, at least for Mysql at least while you are not parsing CHECK expressions. I can still have totally supported i32 and have a CHECK  > 100 in the database level and inserting 12 into the database will happily fail. And diesel can not guaranty that (and in my opinion sure it should not).

So adding support for UNSIGNED DOUBLE  (which is just DOUBLE with CHECK >=0 ) will not break any safety guaranties because there are no such guaranties right now.

The second point is "It's a commonly used type". How common it should be? Do you have some treshhold of number of users?

If you are not planning to provide this support, can you direct me somewhere where I can read about adding support for such fields by myself (my case is not so common but it does not mean me personally is not required to support this case)

 Right now my solution is to manually tweak generated schema every time I make print-schema and remove Unsigned tag

**Reply by weiznich** [reply]
> First of all I want to mention that right now diesel is not fully type safe, at least for Mysql at least while you are not parsing CHECK expressions. I can still have totally supported i32 and have a CHECK > 100 in the database level and inserting 12 into the database will happily fail. And diesel can not guaranty that (and in my opinion sure it should not).

This is currently expected behavior as check constraints are something different than type information. Evaluating `Check` constraints in diesel would require parsing arbitrary SQL and evaluating it. This is for now out of scope.

> So adding support for UNSIGNED DOUBLE (which is just DOUBLE with CHECK >=0 ) will not break any safety guaranties because there are no such guaranties right now.

I know that it is internally implemented that way, but the important difference here is that these information are part of the type instead of being "hidden" in a check constraint. Diesel respects these type information in contrast to the more general `CHECK` constraint, so if something is represented as type it's handled as all the other types.

> The second point is "It's a commonly used type". How common it should be? Do you have some treshhold of number of users?

It's not about the absolute number of users. It's about popularity, so more like which ratio of our users uses that feature. On rust side that means we wont add support for third party types from "small" crates, on sql side that means we won't add support for third party extension types. Having built-in SQL type with a matching rust type in rust-std covers that point sufficiently. As already written above, that's not the reason why I'm not in favor of having built-in support for this in diesel. My problem here is the possible invalid state for the `ToSql` implementation.

> If you are not planning to provide this support, can you direct me somewhere where I can read about adding support for such fields by myself (my case is not so common but it does not mean me personally is not required to support this case)

The [`FromSql`](https://docs.diesel.rs/2.1.x/diesel/deserialize/trait.FromSql.html) and [`ToSql`](https://docs.diesel.rs/2.1.x/diesel/serialize/trait.ToSql.html) documentation shows how to implement support for custom types. 

> Right now my solution is to manually tweak generated schema every time I make print-schema and remove Unsigned tag

You can use a `patch` file to automate that. Checkout the [corresponding guide for some details](http://diesel.rs/guides/configuring-diesel-cli.html)

**Comment by zim32** [comment]
By the way don't whant to look toxic. Want to thank you. Diesel is the most awesome tool I've used. It's compile time checking ability is something new and in general works like a charm.
